File: integration-tests/pkg/cluster/spec.py
from dataclasses import dataclass
from typing import List, Optional
import yaml


@dataclass
class HcloudMachine:
    serverType: str
    datacenter: Optional[str] = None


@dataclass
class Machine:
    id: str
    type: str
    platform: str
    talosInitialVersion: str
    talosImage: str
    privateIP: str
    configPatches: List[str]
    userdata: str
    hcloud: Optional[HcloudMachine] = None


@dataclass
class Cluster:
    name: str
    privateNetwork: str
    privateSubnetwork: str
    kubernetesVersion: str
    machines: List[Machine]


def load(path: str) -> Cluster:
    with open(path, "r", encoding="utf-8") as f:
        data = yaml.safe_load(f)
    machines = [
        Machine(
            configPatches=m.get("configPatches", []),
            userdata=m.get("userdata", ""),
            hcloud=HcloudMachine(**m["hcloud"]) if m.get("hcloud") else None,
            **{
                k: v
                for k, v in m.items()
                if k not in ("configPatches", "userdata", "hcloud")
            }
        )
        for m in data.get("machines", [])
    ]
    return Cluster(
        name=data.get("name", ""),
        privateNetwork=data.get("privateNetwork", ""),
        privateSubnetwork=data.get("privateSubnetwork", ""),
        kubernetesVersion=data.get("kubernetesVersion", ""),
        machines=machines,
    )

File: integration-tests/pkg/cluster/typescript/spec.ts
import { readFileSync } from "fs";

import { parse } from "yaml";

import { validateCluster } from "./validation.js";

import type { Cluster, Machine } from "./cluster.js";

// These specs are internal
type MachineSpec = Omit<Machine, "configPatches" | "applyConfigViaUserdata"> & {
  "apply-config-via-userdata"?: boolean;
  configPatches?: string[];
};

type ClusterSpec = Omit<Cluster, "machines"> & {
  machines: MachineSpec[];
  kubernetesVersion?: string;
  skipInitApply?: boolean;
};

export function load(path: string): Cluster {
  const data = readFileSync(path, "utf8");
  const spec = (parse(data) ?? {}) as Record<string, unknown>;

  validateCluster(spec);
  const typed = spec as ClusterSpec;

  const machines = typed.machines.map((raw) => normalizeMachine(raw));

  return {
    name: typed.name,
    kubernetesVersion: typed.kubernetesVersion ?? "",
    privateNetwork: typed.privateNetwork ?? "",
    privateSubnetwork: typed.privateSubnetwork ?? "",
    skipInitApply: typed.skipInitApply ?? false,
    machines,
  };
}

function normalizeMachine(raw: MachineSpec): Machine {
  const {
    configPatches,
    userdata,
    "apply-config-via-userdata": applyConfigViaUserdataDashed,
    variant,
    ...rest
  } = raw ?? {};

  const applyConfigViaUserdata =
    typeof applyConfigViaUserdataDashed === "boolean"
      ? applyConfigViaUserdataDashed
      : false;

  return {
    ...rest,
    id: raw.id,
    type: raw.type,
    platform: raw.platform,
    talosInitialVersion: raw.talosInitialVersion,
    talosImage: raw.talosImage,
    variant: (variant as Machine["variant"]) || "metal",
    privateIP: raw.privateIP,
    configPatches: Array.isArray(configPatches) ? configPatches : [],
    userdata,
    applyConfigViaUserdata,
    hcloud: raw.hcloud,
  };
}
